src/main.rs

```diff
@@ -10,35 +10,52 @@ use libazureinit::distro::{Distribution, Distributions};
 use libazureinit::{
     error::Error as LibError,
     goalstate, imds, media,
-    media::Media,
+    media::{Environment, Media},
     reqwest::{header, Client},
     user,
 };
 
 const VERSION: &str = env!("CARGO_PKG_VERSION");
 
+// Mount the given device, get OVF environment data, return it.
+fn mount_parse_ovf_env(dev: String) -> Result<Environment, anyhow::Error> {
+    let mount_media =
+        Media::new(PathBuf::from(dev), PathBuf::from(media::PATH_MOUNT_POINT));
+    let mounted = mount_media
+        .mount()
+        .with_context(|| "Failed to mount media.")?;
+
+    let ovf_body = mounted.read_ovf_env_to_string()?;
+    let environment = media::parse_ovf_env(ovf_body.as_str())?;
+
+    mounted
+        .unmount()
+        .with_context(|| "Failed to remove media.")?;
+
+    Ok(environment)
+}
+
 fn get_username(imds_body: String) -> Result<String, anyhow::Error> {
     if imds::is_password_authentication_disabled(&imds_body)? {
         // password authentication is disabled
         Ok(imds::get_username(imds_body.clone())?)
     } else {
         // password authentication is enabled
-        let mount_media = Media::new(
-            PathBuf::from(media::PATH_MOUNT_DEVICE),
-            PathBuf::from(media::PATH_MOUNT_POINT),
-        );
-        let mounted = mount_media
-            .mount()
-            .with_context(|| "Failed to mount media.")?;
 
-        let ovf_body = mounted.read_ovf_env_to_string()?;
-        let environment = media::parse_ovf_env(ovf_body.as_str())?;
+        // list of CDROM devices that is available with possible filesystems.
+        let ovf_devices = media::get_mount_device()?;
+        let mut environment: Option<Environment> = None;
 
-        mounted
-            .unmount()
-            .with_context(|| "Failed to remove media.")?;
+        // loop until it finds a correct device.
+        for dev in ovf_devices {
+            environment = match mount_parse_ovf_env(dev) {
+                Ok(env) => Some(env),
+                Err(_) => continue,
+            }
+        }
 
         Ok(environment
+            .ok_or(libazureinit::error::Error::BlockUtilsError)?
```

Comment:
This error immediately gets turned into `anyhow::Error` (from the return type). Using a library's error outside of the library is somewhat unusual, and particularly in applications you're usually not matching on errors to handle them differently higher up the stack, you're just adding error context and exiting.

You can achieve the same thing by using  `anyhow::anyhow!("Unable to get list of block devices")` here and dropping the `BlockUtilsError` from the library enum:

```suggestion
            .ok_or_else(|| anyhow::anyhow!("Unable to get list of block devices"))?
```

Using `ok_or_else` here will switch from eager to lazy evaluation so you don't create an error if it isn't necessary.